llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -4004,39 +4004,25 @@ KnownBits SelectionDAG::computeKnownBits(SDValue Op, const APInt &DemandedElts,
         }
       }
     } else if (Op.getResNo() == 0) {
-      KnownBits Known0(!LD->getMemoryVT().isScalableVT()
-                           ? LD->getMemoryVT().getFixedSizeInBits()
-                           : BitWidth);
-      EVT VT = Op.getValueType();
-      // Fill in any known bits from range information. There are 3 types being
-      // used. The results VT (same vector elt size as BitWidth), the loaded
-      // MemoryVT (which may or may not be vector) and the range VTs original
-      // type. The range matadata needs the full range (i.e
-      // MemoryVT().getSizeInBits()), which is truncated to the correct elt size
-      // if it is know. These are then extended to the original VT sizes below.
+      unsigned ScalarMemorySize = LD->getMemoryVT().getScalarSizeInBits();
+      KnownBits KnownScalarMemory(ScalarMemorySize);
       if (const MDNode *MD = LD->getRanges()) {
-        computeKnownBitsFromRangeMetadata(*MD, Known0);
-        if (VT.isVector()) {
-          // Handle truncation to the first demanded element.
-          // TODO: Figure out which demanded elements are covered
-          if (DemandedElts != 1 || !getDataLayout().isLittleEndian())
-            break;
-          Known0 = Known0.trunc(BitWidth);
-        }
-      }
+        ConstantInt *Lower = mdconst::extract<ConstantInt>(MD->getOperand(0));
+        KnownBits KnownMetadata(Lower->getBitWidth());
 
-      if (LD->getMemoryVT().isVector())
-        Known0 = Known0.trunc(LD->getMemoryVT().getScalarSizeInBits());
+        computeKnownBitsFromRangeMetadata(*MD, KnownMetadata);
+        KnownScalarMemory = KnownMetadata.anyextOrTrunc(ScalarMemorySize);
```

Comment:
This should be an exact match. If it's not there's a bug somewhere. We're probably not clean on enforcing consistent memory types

Reply:
Removed range metadata if type legalization causes a mismatch.  Now testsuite passes without mismatch.  Thus, removed anyextOrTrunc call.